File: Project31ObjectiveC/Project31ObjectiveC/ViewController.m
//
//  ViewController.m
//  Project31ObjectiveC
//
//  Created by Jonathan Paul on 11/27/23.
//

#import "ViewController.h"
#import "WebKit/WebKit.h"

@interface ViewController ()

@property (strong, nonatomic) IBOutlet UITextField *addressBar;
@property (strong, nonatomic) IBOutlet UIStackView *stackView;
@property (weak, nonatomic) WKWebView *activeWebView;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    [self setDefaultTitle];

    UIBarButtonItem *add = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:self action:@selector(addWebView)];
    UIBarButtonItem *delete = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemTrash target:self action:@selector(deleteWebView)];

    self.navigationItem.rightBarButtonItems = @[delete, add];
}

- (void)setDefaultTitle {
    self.title = @"Multibrowser";
}

- (void)addWebView {
    WKWebView *webView = [WKWebView new];
    webView.navigationDelegate = self;

    [self.stackView addArrangedSubview:webView];

    NSURL *url = [NSURL URLWithString:@"https://www.jonathandpaul.com"];

    [webView loadRequest:[NSURLRequest requestWithURL:url]];

    // Highlight the new webview and recognize when it is tapped on by the user
    webView.layer.borderColor = [[UIColor blueColor] CGColor];
    [self selectWebView:webView];

    UITapGestureRecognizer *recognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(webViewTapped:)]; // add a colon for each param in teh selector
    recognizer.delegate = self;
    [webView addGestureRecognizer:recognizer];
}

// Create a border around the selected web view and set it to the activeWebView property
- (void)selectWebView:(WKWebView *)webView {
    for (UIView *view in self.stackView.arrangedSubviews) {
        view.layer.borderWidth = 0;
    }

    self.activeWebView = webView;
    webView.layer.borderWidth = 3;

    [self updateUI:webView];
}

- (void)updateUI:(WKWebView*)webView {
    self.title = webView.title;
    self.addressBar.text = webView.URL.absoluteString;
}


- (void) webViewTapped:(UITapGestureRecognizer *)recognizer {
    WKWebView *webView = (WKWebView*)recognizer.view;
    [self selectWebView:webView];
}

// WKNavigationDelegate method to update the UI when the WKWebView finishes loading
- (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation {
    if (webView == self.activeWebView) {
        [self updateUI:webView];
    }
}

// TODO: Just like being John Malcovich?
// UIGestureRecognizerDelegate method to tell the gesture to be recognized alongside the recognizers built into the WKWebView
- (BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldRecognizeSimultaneouslyWithGestureRecognizer:(UIGestureRecognizer *)otherGestureRecognizer {
    return YES;
}

// UITextFieldDelegate method when the user clicks return on the keyboard
- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    // This is far less safe without the "if let"s
    NSURL *url = [NSURL URLWithString:self.addressBar.text];
    NSURLRequest *urlRequest = [NSURLRequest requestWithURL:url];
    [self.activeWebView loadRequest:urlRequest];

    // Make the keyboard hide
    [textField resignFirstResponder];
    return YES;
}

- (void)deleteWebView {
    // Because we don't have "if let" we can check for nil to be safe
    if (self.activeWebView == nil) {
        return;
    }
    // In Objective-C the index is of type NSUInteger instead of Array<UIView>.Index which is, in Swift, a typealias to Int
    NSUInteger index = [self.stackView.arrangedSubviews indexOfObject:self.activeWebView];
    [self.activeWebView removeFromSuperview];

    if ([self.stackView.arrangedSubviews count] == 0) {
        // go back to our default UI
        [self setDefaultTitle];
    } else {
        // if that was the last webview in the stack, go back one
        if (index == self.stackView.arrangedSubviews.count) {
            index = self.stackView.arrangedSubviews.count - 1;
        }

        // find the web view at the new index and select it
        // This is done in a less safe way in obj-c
        WKWebView *webView = (WKWebView*)self.stackView.arrangedSubviews[index];
        [self selectWebView:webView];
    }
}

// Deprecated in iOS 17.0
// Have reacive view when multitasking is happening on iPad
- (void)traitCollectionDidChange:(UITraitCollection *)previousTraitCollection {
    if (self.traitCollection.horizontalSizeClass == UIUserInterfaceSizeClassCompact) {
        self.stackView.axis = UILayoutConstraintAxisVertical;
    } else {
        self.stackView.axis = UILayoutConstraintAxisHorizontal;
    }
}



@end

File: Project31Swift/Project31Swift/ViewController.swift
//
//  ViewController.swift
//  Project31Swift
//
//  Created by Jonathan Paul on 11/26/23.
//

import UIKit
import WebKit

class ViewController: UIViewController, WKNavigationDelegate, UITextFieldDelegate, UIGestureRecognizerDelegate {

    @IBOutlet var addressBar: UITextField!
    @IBOutlet var stackView: UIStackView!

    weak private var activeWebView: WKWebView?

    override func viewDidLoad() {
        super.viewDidLoad()

        setDefaultTitle()

        let add = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(addWebView))
        let delete = UIBarButtonItem(barButtonSystemItem: .trash, target: self, action: #selector(deleteWebView))
        navigationItem.rightBarButtonItems = [delete, add]
    }

    private func setDefaultTitle() {
        title = "Multibrowser"
    }

    @objc
    private func addWebView() {
        let webView = WKWebView()
        webView.navigationDelegate = self

        stackView.addArrangedSubview(webView)

        let url = URL(string: "https://www.jonathandpaul.com")!
        webView.load(URLRequest(url: url))

        // Highlight the new webview and recognize when it is tapped on by the user
        webView.layer.borderColor = UIColor.blue.cgColor
        selectWebView(webView)

        let recognizer = UITapGestureRecognizer(target: self, action: #selector(webViewTapped))
        recognizer.delegate = self
        webView.addGestureRecognizer(recognizer)
    }

    // Create a border around the selected web view and set it to the activeWebView property
    private func selectWebView(_ webView: WKWebView) {
        for view in stackView.arrangedSubviews {
            view.layer.borderWidth = 0
        }

        activeWebView = webView
        webView.layer.borderWidth = 3
        updateUI(for: webView)
    }

    private func updateUI(for webView: WKWebView) {
        title = webView.title
        addressBar.text = webView.url?.absoluteString ?? ""
    }

    @objc
    private func webViewTapped(_ recognizer: UITapGestureRecognizer) {
        if let selectedWebView = recognizer.view as? WKWebView {
            selectWebView(selectedWebView)
        }
    }

    // WKNavigationDelegate method to update the UI when the WKWebView finishes loading
    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        if webView == activeWebView  {
            updateUI(for: webView)
        }
    }

    // UIGestureRecognizerDelegate method to tell the gesture to be recognized alongside the recognizers built into the WKWebView
    func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer) -> Bool {
        return true
    }

    // UITextFieldDelegate method when the user clicks return on the keyboard
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        if let webView = activeWebView, let address = addressBar.text {
            if let url = URL(string: address) {
                webView.load(URLRequest(url: url))
            }
        }

        // Make the keyboard hide
        textField.resignFirstResponder()
        return true
    }

    @objc
    private func deleteWebView() {
        // safely unwrap out webview
        if let webView = activeWebView {
            if let index = stackView.arrangedSubviews.firstIndex(of: webView) {
                // We found the webview - remove it from the stack view and destroy it
                webView.removeFromSuperview()

                if stackView.arrangedSubviews.count == 0 {
                    // go back to our default UI
                    setDefaultTitle()
                } else {
                    // convert the Index value into an integer
                    var currentIndex = Int(index)

                    // if that was the last webview in the stack, go back one
                    if currentIndex == stackView.arrangedSubviews.count {
                        currentIndex = stackView.arrangedSubviews.count - 1
                    }

                    // find the web view at the new index and select it
                    if let newSelectedWebView = stackView.arrangedSubviews[currentIndex] as? WKWebView {
                        selectWebView(newSelectedWebView)
                    }
                }
            }
        }
    }

    // Deprecated in iOS 17.0
    // Have reacive view when multitasking is happening on iPad
    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        if traitCollection.horizontalSizeClass == .compact {
            stackView.axis = .vertical
        } else {
            stackView.axis = .horizontal
        }
    }

}

